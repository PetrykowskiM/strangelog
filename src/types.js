// @flow

export type ConfigType = {
  path: string,
  components: ComponentsConfigType
};

export type ComponentsConfigType = {
  [name: string]: string
};

export type VersionType = {
  major: number,
  minor: number,
  patch: number
};

export type EntryKindType = 'addition' | 'change' | 'fix';

export type EntryType = {
  component: string,
  kind: EntryKindType,
  description: string
};

export type VersionChangelogType = {
  version: ?VersionType,
  entries: {
    [kind: EntryKindType]: EntryType[]
  }
};

export type ChangelogType = VersionChangelogType[];

export type ChangelogAPIType = {
  addEntry: (
    entry: EntryType
  ) => void;
  bumpNextVersion: (nextVersion: VersionType) => void,
  generate: () => string,
  getChangelogData: () => ChangelogType,
  getPossibleNextVersions: () => VersionType[] | null,
  getComponentsConfig: () => ComponentsConfigType
};

export type CLIOptionsType = {
  _: string[],
  directory: string
};
