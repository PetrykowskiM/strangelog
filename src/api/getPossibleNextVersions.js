// @flow

import type { ConfigType, VersionType } from '../types';

import getSortedChangelogVersions from './getSortedChangelogVersions';

export default function getPossibleNextVersions(
  config: ConfigType
): VersionType[] | null {
  const latestVersion = getLatestChangelogVersion(config);

  if (!latestVersion) {
    return null;
  }

  return [{
    major: latestVersion.major,
    minor: latestVersion.minor,
    patch: latestVersion.patch + 1
  }, {
    major: latestVersion.major,
    minor: latestVersion.minor + 1,
    patch: 0
  }, {
    major: latestVersion.major + 1,
    minor: 0,
    patch: 0
  }];
}


export function getLatestChangelogVersion(config: ConfigType) {
  return getSortedChangelogVersions(config)[0];
}