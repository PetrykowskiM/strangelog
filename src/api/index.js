// @flow

import connectChangelog from './connectChangelog';

export { connectChangelog };
