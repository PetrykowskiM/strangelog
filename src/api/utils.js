// @flow

import type { VersionType } from '../types';

export function stringifyVersion(version: ?VersionType): string {
  if (!version) {
    return 'next';
  }

  const { major, minor } = version;

  return `${major}.${minor}`;
}

export function multiToSingleLineString(multiLineIndentedString: string): string {
  return multiLineIndentedString.replace(/\n[ \t]*/g, ' ').trim();
}
