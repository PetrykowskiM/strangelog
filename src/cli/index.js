#!/usr/bin/env node

// @flow

import cli from './cli';

// See https://github.com/yargs/yargs/issues/605
cli(process.argv.slice(2));
