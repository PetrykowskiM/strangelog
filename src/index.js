// @flow

export * from './api';
export * from './types';