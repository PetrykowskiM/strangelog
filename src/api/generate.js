// @flow

import { join as joinPath } from 'path';
import { readFileSync } from 'fs-extra';
import { create as createHandlebars } from 'handlebars';

import type { ConfigType, ChangelogType } from '../types';

import { stringifyVersion } from './utils';

const handlebarsInstance = createHandlebars();

handlebarsInstance.registerHelper('stringifyVersion', (version) => stringifyVersion(version));

export default function generate(
  { path, components }: ConfigType,
  changelog: ChangelogType
): string {
  const generateMarkdown = handlebarsInstance.compile(
    readFileSync(joinPath(__dirname, '../../templates/CHANGELOG.hbs')).toString()
  );

  return generateMarkdown({
    changelog
  });
}
