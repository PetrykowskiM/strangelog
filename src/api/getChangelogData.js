// @flow

import { join as joinPath } from 'path';

import { readFileSync } from 'fs-extra';
import groupBy from 'lodash/groupBy';
import jsYaml from 'js-yaml';
import { sync as syncGlob } from 'glob';

import type {
  ConfigType,
  ChangelogType,
  VersionChangelogType,
  VersionType
} from '../types';

import { stringifyVersion } from './utils';
import getSortedChangelogVersions from './getSortedChangelogVersions';


export default function getChangelogData(
  config: ConfigType
): ChangelogType {
  return [
    getVersionChangelog(config, null),
    ...getSortedChangelogVersions(config).map((version) =>
      getVersionChangelog(config, version))
  ];
}

function getVersionChangelog(
  config: ConfigType,
  version: ?VersionType
): VersionChangelogType {
  return {
    version,
    entries: {
      addition: [],
      change: [],
      fix: [],
      ...groupBy((
        getVersionChangelogFileNames(config, stringifyVersion(version))
          .map((entryFileName) => readFileSync(entryFileName).toString())
          .map(jsYaml.safeLoad)
      ), 'kind')
    }
  };
}

function getVersionChangelogFileNames(
  { path }: ConfigType,
  versionString: string
): string[] {
  return syncGlob(joinPath(path, versionString, '*.yml'));
}