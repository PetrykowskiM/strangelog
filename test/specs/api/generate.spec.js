// @flow

import { removeSync } from 'fs-extra';

import { addTestVersionsWithEntries } from '../../factories/changelog';
import { connectChangelog } from '../../../src/api';
import { getOwnTestPath } from '../../factories/fileSystem';

const testPath = getOwnTestPath();

describe('generate', () => {

  beforeEach(() => {
    removeSync(testPath);
  });

  afterEach(() => {
    removeSync(testPath);
  });

  function setup() {
    return connectChangelog({
      path: testPath,
      components: {
        comp1: 'Comp 1',
        comp2: 'Comp 2'
      }
    });
  }

  it('returns proper markdown string', () => {
    const changelogAPI = setup();

    addTestVersionsWithEntries(changelogAPI);

    expect(
      changelogAPI.generate()
    ).toMatchSnapshot();
  });

});
