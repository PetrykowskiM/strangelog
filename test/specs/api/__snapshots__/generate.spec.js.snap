// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate returns proper markdown string 1`] = `
"# Web Frontend Changelog

## Version \`next\`

### Added
_No entries_

### Changed
_No entries_

### Fixed
- **Comp 2:** comp2 fix description

## Version \`1.1.0\`

### Added
_No entries_

### Changed
- **Comp 1:** comp1 change description

### Fixed
_No entries_

## Version \`1.0.0\`

### Added
- **Comp 1:** comp1 addition description

### Changed
_No entries_

### Fixed
_No entries_

"
`;
